Font size / page size setting for printing

carlobeltrame opened the pull request:
Fixes #4650 
<img width="1245" height="729" alt="image" src="https://github.com/user-attachments/assets/b6264b6d-833b-4737-b42c-afa1c38e3907" />

A5 / A6 (bigger relative font size): [Nuxt A5](https://github.com/user-attachments/files/21986312/Pfila-2023-185.pdf) [Client A5](https://github.com/user-attachments/files/21986288/Pfila-2023-184.pdf)
A4 / A5 (normal, same as on prod): [Nuxt A4](https://github.com/user-attachments/files/21986330/Pfila-2023-186.pdf) [Client A4](https://github.com/user-attachments/files/21986335/Pfila-2023-187.pdf)
~~A4 / A3 (smaller relative font size): [Nuxt A3](https://github.com/user-attachments/files/21986344/Pfila-2023-188.pdf) [Client A3](https://github.com/user-attachments/files/21986346/Pfila-2023-189.pdf)~~

Please feedback on:
- UI design of the new option
- Currently, we use A4 page sizes but many users print this on A5. I took the easy way out and left the page size default at A4 (so I don't have to change all font sizes in both print) and just labeled it as "Normal (for printing on A4 or A5)" in the UI. So when the user selects the font size "Bigger (for printing on A6)", the PDF actually uses a page size of A5, and we rely on the users' printers to scale this down. Is this fine? Do users think more in terms of page size or in terms of font size? Should I do the tedious work and change all font sizes in both prints, so that we can be honest to the users?
- Any bugs in the pdfs. I already had to adjust the picasso time column font size in nuxt print and the auto-scaling font size on picasso schedule entries in client print already.

carlobeltrame committed:
Simple font size implementation, fixes #4650

BacLuc commented:
The space problem for small blocks get worse on the picasso:

Nuxt A5:
<img width="1826" height="1226" alt="image" src="https://github.com/user-attachments/assets/f812642d-d32e-466f-9d2e-6d417a90eb4f" />

And the 07:00 is over the line.

If for whatever reason you use 2 zeros for the relative time from block start, it always wraps:
<img width="1826" height="1226" alt="image" src="https://github.com/user-attachments/assets/d18401d3-684f-40a5-a8dc-5f0efcb4a3fd" />
I didn't write many blocks who were longer than 10h...?
But the problem persist for people who use absolute times.

Here it is completely off:
<img width="1246" height="1072" alt="image" src="https://github.com/user-attachments/assets/7e55a325-da77-4de0-85ce-95a3267f5509" />

Material list looks good:
<img width="1266" height="630" alt="image" src="https://github.com/user-attachments/assets/c9adce68-2d17-4e02-9b1d-0592c82971af" />

ThematicArea looks good:
<img width="1220" height="332" alt="image" src="https://github.com/user-attachments/assets/cd9b6cc3-1ed7-464d-88af-a29189ab599e" />

[Client A5](https://github.com/user-attachments/files/21986288/Pfila-2023-184.pdf)
Picasso looks good:
<img width="2972" height="2080" alt="image" src="https://github.com/user-attachments/assets/3a744c3d-905d-4ba7-9e8c-479c43a15dc3" />

Here the user can use 2 digit hours in the block:
<img width="2972" height="2080" alt="image" src="https://github.com/user-attachments/assets/f9a40954-0b2c-4e63-8fb1-56f83bef9a80" />

A3:

I don't know what you would want to print on A3, maybe the picasso?
This works, looks a little empty.

UI Design:
Good that it's hidden and does not clutter the ui for not advanced users.
Not so discoverable...but you have to weigh the 2 options.

BacLuc commented on frontend/src/components/print/__tests__/repairPrintConfig.spec.js:
```diff
@@ -106,6 +106,10 @@ describe('repairConfig', () => {
       },
     },
   ]
+  const defaultOptions = {
```
❤️

BacLuc commented on frontend/src/components/print/__tests__/repairPrintConfig.spec.js:
```diff
@@ -396,11 +400,281 @@ describe('repairConfig', () => {
+  })
+
+  const pageSizes = [('A5', 'A4', 'A3')]
+  pageSizes.forEach((pageSize) => {
```
Cool, nitpick:
With vitest you have to get used to it, but the syntax for parametrized tests is:

```
describe('idToColor', () => {
  it.each([
    [[undefined, false], '#4d4d4d'],
    [[null, false], '#4d4d4d'],
    [['', false], '#4d4d4d'],
    [['0000000', false], '#900'],
    [['0000000', true], '#4d4d4d'],
    [['fffffff', false], '#992600'],
    [['Wrong input', false], '#900'],
  ])('maps %p to %p', (input, expected) => {
```

See https://github.com/ecamp/ecamp3/blob/9e921c611dd87e7a0d2aa081d572d0317bb85ef3/common/helpers/__tests__/colors.spec.js#L5 as on example.
Or search "it.each"

carlobeltrame replied:
Yeah I've been in the ruby testing world a lot... e4d3c1f

BacLuc commented on frontend/src/components/print/__tests__/repairPrintConfig.spec.js:
```diff
@@ -396,11 +400,281 @@ describe('repairConfig', () => {
+      })
+    })
+  })
+
```
Maybe add: Repairs invalid page size

carlobeltrame replied:
Good idea! Done in e4d3c1f

BacLuc commented on pdf/src/campPrint/picasso/ScheduleEntry.vue:
```diff
@@ -50,8 +51,15 @@ export default {
       return `#scheduleEntry_${this.scheduleEntry.id}`
     },
+    fontSizeScalingFactor() {
+      return {
```
Is this a different way for a switch statement?

carlobeltrame replied:
Replaced in e4d3c1f

BacLuc commented on print/components/activityList/ActivityListPeriod.vue:
```diff
@@ -36,6 +36,7 @@ const props = defineProps({
   },
   index: { type: Number, required: true },
   filter: { type: Object, default: () => ({}) },
+  pageSize: { type: String, default: 'a4' },
```
nitpick:
In clientprint its 'A4', here 'a4'

carlobeltrame replied:
That's actually intentional. In nuxt print, `a4` is used as a CSS class name, and those are usually lowercase. In client print, `A4` needs to be passed into the `<Page>` element, and it expects it to be uppercase.

BacLuc commented:
What we also have to discuss:

How much value do we get vs additional maintenance effort?

Do we need tests for the other font sizes?
How do we test the other font sizes?

How much priority does it get when it breaks?

carlobeltrame commented:
> The space problem for small blocks get worse on the picasso

True, but if you are using this advanced feature, you probably can also set the picasso to portrait mode and maybe even split it in two parts using the date filters, so it's less of a problem.

> And the 07:00 is over the line.

Fixed in 10724c5

> I don't know what you would want to print on A3, maybe the picasso?
> This works, looks a little empty.

> How much value do we get vs additional maintenance effort?

I could also imagine dropping the A3 option, and just supporting the A6 one for now. I agree A4/A3 is only useful for the picasso, if at all. It does improve the "space problem for small blocks" which you mention on the A6 version, but nobody has explicitely asked for reducing the font size as far as I am aware (although when users claim they can create a better readable picasso in excel, this usually just boils down to a smaller font size).

I implemented this because I see the A6 variant as a valid use case for which we can't provide any workarounds at the moment. And if it's just setting the page size plus 1-2 workarounds for the smaller paper size, I feel like the maintenance burden is not so high.

> Do we need tests for the other font sizes? How do we test the other font sizes?

~~I now added basic e2e tests just checking for the number of pages in the A6 variant. 28461de~~ The e2e tests turned out to be extremely flaky in firefox for some reason... I removed them again and added snapshot tests in the pdf module (9c6cda3). In print, we don't have any logic tests at all so far...

carlobeltrame committed:
Add tests for different page sizes

carlobeltrame committed:
Don't support A3 for now

carlobeltrame committed:
Adjust existing tests

manuelmeister commented:
Grössere Schrift (optimiert für A6)
Normale Schrift (optimiert für A4 oder A5)

carlobeltrame committed:
Adjust translations
